### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by devendra-alt**
### 👨‍💻 Requirments fulfilled
✔️ Added classroom class
✔️ Added book class
✔️ Added rental class
✔️ Added `relationship` between classes
✔️ Added `to-many` and `belong-to` relationship

### CODE REVIEW

File: school/classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  attr_reader :students
+
+  def initialize(label)
+    @label = label
+    @students = []
+  end
+
+  def add_student(student)
+    @student.push(student)
+  end
```
Comment (TedLivist) [inline school/classroom.rb]: - [x] What you declared is a `@students` array not `@student`. Kindly fix this error.

- [x] Apart from adding the student to the array you still need to set the classroom of the student to this classroom to complete the association.

  ```
  def add_student(student)
    @students.push(student)
    student.classroom = self
  end
  ```

File: school/student.rb [hunk 2]
```diff
@@ -1,5 +1,7 @@
 require_relative 'person'
 class Student < Person
```
Comment (TedLivist) [inline school/student.rb:2]: - [x] Kindly set up the other end of the classroom association in this student class by adding its method.

  ```
  def classroom=(classroom)
    @classroom = classroom
    classroom.students.push(self) unless classroom.students.include?(self)
  end
  ```

File: school/rental.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Rental
+  attr_accessor :date, :book
```
Comment (TedLivist) [inline school/rental.rb]: - [x] Kindly add the `:person` attribute to this getter and setter

  ```
  attr_accessor :date, :book, :person
  ```

File: school/rental.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+
+  def initialize(date, book, person)
+    @date = date
+    @book = book
+    @person = person
+  end
```
Comment (TedLivist) [inline school/rental.rb]: - [x] After creating this rental we need to add it to the respective rental arrays of the book and person, as a rental cannot exist independent of the two objects.

  ```
  def initialize(date, book, person)
    @date = date
    
    @book = book
    book.rentals << self
    
    @person = person
    person.rentals << self
  end
  ```

File: school/person.rb [hunk 5]
```diff
@@ -19,6 +20,10 @@ def of_age?
 
   public
 
+  def add_rental(rental)
+    @rental = rental
+  end
```
Comment (TedLivist) [inline school/person.rb]: - [x] The purpose of the `add_rental` method is to allow the user to create a rental given a book and a date, so this method should be more about the instantiation of the rental class.

  ```
  def add_rental(date, book)
    Rental.new(date, book, self)
  end
  ```

File: school/book.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals.push(rental)
+  end
```
Comment (TedLivist) [inline school/book.rb]: - [x] The purpose of the `add_rental` method is to allow the user to create a rental given a person and a date, so this method should be more about the instantiation of the rental class.

  ```
  def add_rental(date, person)
    Rental.new(date, self, person)
  end
  ```

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @devendra-alt,

Good job so far!
✔️ Implementation of association methods
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by prolajumokeoni** [comment]
Hi @devendra-alt, please fix the Rubocop linter errors on your project 👍🏿

**Review by julie-ify — COMMENTED** [review]
Hi @devendra-alt,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

* Kindly verify your submissions before requesting for reviews. This is a set up association project but you requested a review using the basic UI form. Kindly pay attention your submissions.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: school/person.rb [hunk 7]
```diff
@@ -19,6 +20,10 @@ def of_age?
 
   public
 
+  def add_rental(date, book)
+    Rental.new(date, book)
+  end
```
Comment (tufoinnkuo10) [inline school/person.rb]: - [x] Great job so far. Please, like the previous reviewer noted, The person itself should be sent as a parameter to create the new rental, achieving the 'has-many' association. Kindly include `self `like the previous review précised.
ie
```ruby
 def add_rental(book, date)
    Rental.new(date, book, self)
  end
```

**Review by tufoinnkuo10 — CHANGES_REQUESTED** [review]
Hi @devendra-alt 

great job making the changes. let's perfect it before you go to the next project but you are almost there!

To Highlight
✅ You used the correct Gitflow
✅ Descriptive pull request message
✅ linter error checks validated
✅ associations well set


Required Changes ♻️
Check the comments under the review
Optional suggestions
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)

**Review by mckent05 — APPROVED** [review]
Hi @devendra-alt, 

# STATUS: APPROVED ✔️ ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations and well done! 🎉 
approved


## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👏 👏 
- Good job creating associations between your classes 👍 👍 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**